tests/unit/websocket_test.cc

```diff
@@ -64,3 +69,60 @@ SEASTAR_TEST_CASE(test_websocket_handshake) {
         }
     });
 }
+
+SEASTAR_TEST_CASE(test_websocket_handler_registration) {
+    return seastar::async([] {
+        loopback_connection_factory factory;
+        loopback_socket_impl lsi(factory);
+
+        auto acceptor = factory.get_server_socket().accept();
+        auto connector = lsi.connect(socket_address(), socket_address());
+        connected_socket sock = connector.get0();
+        auto input = sock.input();
+        auto output = sock.output();
+
+        // Setup server
+        websocket::server ws;
+        ws.register_handler("echo", [] (input_stream<char>& in,
+                    output_stream<char>& out) {
+            // TODO: Fix along with echo handler in demo.
+            return out.write("TEST");
+        });
+        websocket::connection conn(ws, acceptor.get0().connection);
+        future<> serve = conn.process();
+
+        auto close = defer([&conn, &input, &output, &serve] () noexcept {
+            conn.shutdown();
```

Comment:
... and here, call shutdown() just like before, but just add a call to `conn.close().get()`. The reason for the split is that otherwise it's only legal to call `shutdown` from a seastar thread (seastar::async), and we should avoid such restrictions if possible. After the split, it would also be possible to close the connection from code that is not inside seastar::async context.